searchlib/src/vespa/searchlib/query/query_term_ucs4.cpp

```diff
@@ -43,6 +36,25 @@ QueryTermUCS4::QueryTermUCS4(const string & termS, Type type) :
     }
 }
 
+void
+QueryTermUCS4::fillUCS4() {
+    /*
+     * Double checked locking......
+     * This is a 'dirty' optimisation, but this is done to avoid writing a lot of data and blow the cpu caches with something
+     * you do not really need most of the time. That matters when qps is very high and query is wide, and hits are few.
+     */
+    std::lock_guard guard(_globalMutex);
+    if (_filled) return;
+    _termUCS4.reset(new ucs4_t[_cachedTermLen + 1]);
+    vespalib::Utf8Reader r(getTermString());
+    uint32_t i(0);
+    while (r.hasMore()) {
+        _termUCS4.get()[i++] = r.getChar();
+    }
+    _termUCS4.get()[_cachedTermLen] = 0;
+    _filled = true;
```

Comment:
Might need a write barrier before setting _filled ?

Reply:
I guess you get a barrier when assigning an atomic bool. I think that is default.